rayon-core/src/registry.rs

```diff
@@ -479,26 +480,34 @@ impl Registry {
             }
         }
     }
-
+    
+    
     #[cold]
     unsafe fn in_worker_cold<OP, R>(&self, op: OP) -> R
     where
         OP: FnOnce(&WorkerThread, bool) -> R + Send,
         R: Send,
     {
-        // This thread isn't a member of *any* thread pool, so just block.
-        debug_assert!(WorkerThread::current().is_null());
-        let job = StackJob::new(
-            |injected| {
-                let worker_thread = WorkerThread::current();
-                assert!(injected && !worker_thread.is_null());
-                op(&*worker_thread, true)
-            },
-            LockLatch::new(),
-        );
-        self.inject(&[job.as_job_ref()]);
-        job.latch.wait();
-        job.into_result()
+        LOCK_LATCH.with(|l| {
+            // This should not panic since the latch is thread local and we are in the `cold` path.
+            // If `op` were to call another `ThreadPool::install` it would not end up here.
+            let latch = l.borrow();
+
+            // This thread isn't a member of *any* thread pool, so just block.
+            debug_assert!(WorkerThread::current().is_null());
+            let job = StackJob::new(
+                |injected| {
+                    let worker_thread = WorkerThread::current();
+                    assert!(injected && !worker_thread.is_null());
+                    op(&*worker_thread, true)
+                },
+                latch.deref(),
+            );
+            self.inject(&[job.as_job_ref()]);
+            job.latch.wait();
+            job.latch.reset(); // Makes sure we can use the same latch again next time.
```

Comment:
Can we make this a unified `wait_and_reset` instead? That way it doesn't have to be re-locked for the reset.